module/zfs/qat_crypt.c

```diff
@@ -0,0 +1,387 @@
+	if (status != CPA_STATUS_SUCCESS)
+		goto error;
+	session_init = CPA_TRUE;
+
+	nr_bufs = enc_len / PAGE_CACHE_SIZE +
+	    (enc_len % PAGE_CACHE_SIZE == 0 ? 0 : 1);
+	status = init_cy_buffer_lists(cy_inst_handle, nr_bufs, &src_buffer_list,
+	    &dst_buffer_list);
+	if (status != CPA_STATUS_SUCCESS)
+		goto error;
+
+	status = PHYS_CONTIG_ALLOC(&flat_src_buf_array,
+	    nr_bufs * sizeof (CpaFlatBuffer));
+	if (status != CPA_STATUS_SUCCESS)
+		goto error;
+	status = PHYS_CONTIG_ALLOC(&flat_dst_buf_array,
+	    nr_bufs * sizeof (CpaFlatBuffer));
+	if (status != CPA_STATUS_SUCCESS)
+		goto error;
+
+	bytes_left = enc_len;
+	in = src_buf;
+	out = dst_buf;
+	flat_src_buf = flat_src_buf_array;
+	flat_dst_buf = flat_dst_buf_array;
+	while (bytes_left > 0) {
+		in_pages[page_num] = mem_to_page(in);
+		out_pages[page_num] = mem_to_page(out);
+		flat_src_buf->pData = kmap(in_pages[page_num]);
+		flat_dst_buf->pData = kmap(out_pages[page_num]);
+		flat_src_buf->dataLenInBytes = min((long)PAGE_CACHE_SIZE,
+		    (long)bytes_left);
+		flat_dst_buf->dataLenInBytes = min((long)PAGE_CACHE_SIZE,
+		    (long)bytes_left);
+		in += flat_src_buf->dataLenInBytes;
+		out += flat_dst_buf->dataLenInBytes;
+		bytes_left -= flat_src_buf->dataLenInBytes;
+		flat_src_buf++;
+		flat_dst_buf++;
+		page_num++;
+	}
+	src_buffer_list.pBuffers = flat_src_buf_array;
+	dst_buffer_list.pBuffers = flat_dst_buf_array;
+
+	op_data.sessionCtx = cy_session_ctx;
+	op_data.packetType = CPA_CY_SYM_PACKET_TYPE_FULL;
+	op_data.pIv = NULL; /* set this later as the J0 block */
+	op_data.ivLenInBytes = 0;
+	op_data.cryptoStartSrcOffsetInBytes = 0;
+	op_data.messageLenToCipherInBytes = 0;
+	op_data.hashStartSrcOffsetInBytes = 0;
+	op_data.messageLenToHashInBytes = 0;
+	op_data.pDigestResult = 0;
+	op_data.messageLenToCipherInBytes = enc_len;
+	op_data.ivLenInBytes = ZIO_DATA_IV_LEN;
+	op_data.pDigestResult = digest_buf;
+	op_data.pAdditionalAuthData = aad_buf;
+	op_data.pIv = iv_buf;
+
+	cb.verify_result = CPA_FALSE;
+	init_completion(&cb.complete);
+	status = cpaCySymPerformOp(cy_inst_handle, &cb, &op_data,
+	    &src_buffer_list, &dst_buffer_list, NULL);
+	if (status != CPA_STATUS_SUCCESS)
+		goto error;
+
+	if (!wait_for_completion_interruptible_timeout(&cb.complete,
+	    QAT_TIMEOUT_MS)) {
+		status = CPA_STATUS_FAIL;
+		goto error;
+	}
+
+	if (cb.verify_result == CPA_FALSE) {
+		status = CPA_STATUS_FAIL;
+		goto error;
+	}
+
+	if (dir == QAT_ENCRYPT)
+		QAT_STAT_INCR(encrypt_total_out_bytes, enc_len);
+	else
+		QAT_STAT_INCR(decrypt_total_out_bytes, enc_len);
+
+	for (i = 0; i < page_num; i ++) {
+		kunmap(in_pages[i]);
+		kunmap(out_pages[i]);
+	}
+
+	cpaCySymRemoveSession(cy_inst_handle, cy_session_ctx);
+	PHYS_CONTIG_FREE(src_buffer_list.pPrivateMetaData);
+	PHYS_CONTIG_FREE(dst_buffer_list.pPrivateMetaData);
+	PHYS_CONTIG_FREE(cy_session_ctx);
+	PHYS_CONTIG_FREE(flat_src_buf_array);
+	PHYS_CONTIG_FREE(flat_dst_buf_array);
+
+	return (CPA_STATUS_SUCCESS);
+
+error:
```

Comment:
Let's model this after `qat_compress()` which shares a common cleanup path for success and failure.  Lines 351-363 are virtually identical to 370-383.